### DISCUSSION TITLE
PiPedal v1.2.51 Release

### DISCUSSION THREAD

**Discussion by rerdavies** [rerdavies/pipedal#200] [Releases]
Click here to download [PiPedal v1.2.51 Release](https://rerdavies.github.io/pipedal/download.html)

<b>Important notice:</b> Changes to the Auto-update code prevent previous versions from automatically updating to this version. You will have to install this version by downloading the [debian package](https://rerdavies.github.io/pipedal/download.html) and installing it manually.

This release replaces Wi-Fi Direct connections with Wi-Fi hotspots. Support for Wi-Fi Direct on Linux and Android has been fragile for some time. An update to Raspberry Pi OS in early September broke Wi-Fi Direct support completely. As it turns out, Auto-hotspots work much better. 

The new Auto-Hostpot feature in PiPedal allows you to configure your Raspberry Pi so it automatically starts a Wi-Fi hotspot whenever you are away from home. The updated Android PiPedal Client will automatically detect and connect to your Raspberry PI whenever it is visible on the current Wi-Fi network. An updated Android [PiPedal Remote](https://play.google.com/store/apps/details?id=com.twoplay.pipedal&hl=en_US&pli=1) app has been posted on Google Play. Make sure you are using the updated version.

New MIDI system bindings allow you to enable or disable the Wi-Fi hotspot, and to shut down or reboot your Raspberry Pi using MIDI-triggered events.

NOTICE: PiPedal 1.2.47 fixed a significant and dangerous defect that may cause loss of presets, banks or configuration data if you remove power from your Raspberry Pi within up to five minutes of last saving data instead of performing an orderly shutdown or reboot. Users of PiPedal should upgrade to at least version 1.2.47 immediately, if they have not already done so.

Bug fixes:
- PiPedal Remote client theme native pages do not match the selected PiPedal UI theme.
- Unable to configure audio buffers properly for devices with minBufferSize=256.
- Release and Beta updates not visible if the most recent release is a development build.
- Revised auto-update procedure, including a new signing procedure.
- Ongoing improvements to TooB Neural Amp Modeler (bug fixes, performance improvements)
- pipedald service hangs and/or throws exceptions on shutdown.
- Update DNS/SD service announcements when the device name is changed. 
- Unannounce DNS/DS services when the pipedal services shuts down.
- Auto-uprade checks may exceed the throttling rate of Github API calls when checking for updates.
- Memory corruption when ALSA device creation fails.
- MIDI input causing audio glitches and underruns.
- Unable to connect when using IPv6 connections and mDNS name resolution.
- Default web port configuration overwritten by upgrades.
- Remove dependencies on old pre-NetworkManager network services required by Wifi-Direct code.
- Refuse to install on versions of Raspberry Pi OS prior to bookworm.

Known issues:
- TooB Neural Amp Modeler runs 5% slower than Mike Oliphant's Neural Amp Modeler, despite running virtually identical code from Steven Atkins' NeuralAmpModelerCore project. Investigation continues.

**Comment by 38github** [comment]
> TooB Neural Amp Modeler runs 5% slower than Mike Oliphant's Neural Amp Modeler, despite running virtually identical code from Steven Atkins' NeuralAmpModelerCore project. Investigation continues.

Have you compiled it with Clang? That makes a difference compared to GCC. Also -O3 and -Ofast makes a difference since, if I understand it correctly, it vectorizes more instructions.

**Comment by rerdavies** [comment]
I have -O3 and -ofast. And the code all vectorizes properly. Clang is
usually slightly slower than GCC; but that could account for the
difference. That would certainly explain why the profiling results look
weirdly different even though both plugins are running substantially the
same code.I'll give that a try. Thanks.

**Comment by 38github** [comment]
Was the 8% CPU gain from using Clang? If I remember correctly the LSTM models will gain even larger performance by compiling with -fopenmp. Also a tip if you use neural-amp-modeler-lv2 code is to edit the dcblock part because its high-pass effect is way to much. I change the 220.0 value to just 1.0 on row 291 in neural-amp-modeler-lv2/tree/main/src/nam_plugin.cpp

**Comment by rerdavies** [comment]
Clang didn't actually didn't give me a boost. It ended up being a very
subtle code tweak that I don't honestly fully understand.

Toob NAM uses a different DC blocker. I can't remember what the cutoff is,
but it's low. 10hz? (Note to self to check that it is). And the matrices
involved are probably not big enough to benefit from openmp.

Soooo close to being able to use two instances in a single patch on a Pi 4
:p

**Reply by 38github** [reply]
Two instances of STANDARD models on an RPi4 is good! I think I can do that on my RPi4 (neural-amp-modeler-lv2) if I use no other effects but I will have to try it out. I use one FEATHER (pedals) and one STANDARD (amplifier) plus Cab IRs, pre and post EQs, delay, expander, etc and it uses ~80% at 64/4 latency.
